buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java

```diff
@@ -0,0 +1,229 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.api.Buildable;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.TaskDependency;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.Locale;
+
+public class ElasticsearchDistribution implements Buildable {
+
+    public enum Platform {
+        LINUX,
+        WINDOWS,
+        DARWIN;
+
+        @Override
+        public String toString() {
+            return super.toString().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    public enum Type {
+        INTEG_TEST_ZIP,
+        ARCHIVE,
+        RPM,
+        DEB;
+
+        @Override
+        public String toString() {
+            return super.toString().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    public enum Flavor {
+        DEFAULT,
+        OSS;
+
+        @Override
+        public String toString() {
+            return super.toString().toLowerCase(Locale.ROOT);
+        }
+    }
+
+    // package private to tests can use
+    static final Platform CURRENT_PLATFORM = OS.<Platform>conditional()
+        .onLinux(() -> Platform.LINUX)
+        .onWindows(() -> Platform.WINDOWS)
+        .onMac(() -> Platform.DARWIN)
+        .supply();
+
+    public static final class Extracted implements Buildable, Iterable<File> {
+
+        // pkg private so plugin can configure
+        final Configuration configuration;
+
+        private Extracted(Configuration configuration) {
+            this.configuration = configuration;
+        }
+
+        @Override
+        public Iterator<File> iterator() {
+            return configuration.iterator();
+        }
+
+        @Override
+        public TaskDependency getBuildDependencies() {
+            return configuration.getBuildDependencies();
+        }
+
+        @Override
+        public String toString() {
+            return configuration.getSingleFile().toString();
+        }
+    }
+
+    private final String name;
+    // pkg private so plugin can configure
+    final Configuration configuration;
+    private final Extracted extracted;
+
+    private final Property<Version> version;
+    private final Property<Type> type;
+    private final Property<Platform> platform;
+    private final Property<Flavor> flavor;
+    private final Property<Boolean> bundledJdk;
+
+    ElasticsearchDistribution(String name, Project project) {
+        this.name = name;
+        this.configuration = project.getConfigurations().create("es_distro_file_" + name);
```

Comment:
Can we avoid mutating the Gradle model in constructors ?

Reply:
Agreed. This is not best practice. We should probably create the configuration in the plugin and pass as a constructor argument. We could also then avoid passing `Project` altogether and instead inject an `ObjectFactory` for all the property creation.